### PR TITLE
Added extraction from html like structures

### PR DESCRIPTION
**Pull request by blake-newman**
- Added appropriate tests
- Ensured mapping of lines and indentation remains
- Works with direct string input
- Works with files
- Added CLI flag
- Added documentation

### CODE REVIEW

**Comment by davidtheclark** [comment]
Whoah. This is unexpected and awesome!

I will have to look at this slowly sometime soon to make sure I understand all that's going on, hopefully within the next few days.

In the meanwhile, you could fix the linting error and get the CI passing. Also, did you have any ideas about documenting this feature?

**Comment by blake-newman** [comment]
Yeah can help create the docs, what would be easier would to enable configuration option to turn it on, to make IDE linters more acceptable to this. This will also mean less work when people don't want it turned on. 

How would you recommend doing this? 

Ill fix that lint issue now.

**Comment by davidtheclark** [comment]
This should definitely should be an option to avoid unnecessary overhead for the majority of use-cases.

I wonder if this should be a `syntax` option (alonside `scss`, `less`, and `sugarss`): `html` could mean parse HTML and run the `<style>` tags' content through stylelint (as regular CSS, of course).

Any other ideas for option names/places?

I do wonder if the bulk of this should be or use a separate module. The task of extracting CSS from HTML is pretty generic, definitely not stylelint-specific; so I'd imagine there are implementations available on npm, or else there should be. Why don't we use & contribute to one of those? Do you want to look into that?

**Comment by blake-newman** [comment]
@davidtheclark Problem with using the syntax option is that you may have a style tag using scss/postcss as it's syntax. Especially with Vue files. 

I did attempt to find a module to do this, however inorder to keep the mapping of lines and indentation, it needs to be done like so. Other modules will only extract the css making it impossible to map the numbers and indentation.

**Comment by blake-newman** [comment]
I belive to have custom options passed, we will need consolidate the config, configOverride ect to get the correct state. As you may want to always force your IDE to lint html/vue files so this could be an option with the linters and the IDE.

**Comment by davidtheclark** [comment]
> you may have a style tag using scss/postcss as it's syntax. Especially with Vue files.

Did not know that.

Would `extractStyleTagsFromHtml` be a good option name? Or too verbose?

`extractFromHtml`?

File: src/__tests__/standalone-test.js [hunk 1]
```diff
@@ -166,6 +166,27 @@ test("standalone with input css and quiet mode", t => {
   t.plan(planned)
 })
 
+test("standalone with extration from html", t => {
```
Comment (davidtheclark) [inline src/__tests__/standalone-test.js]: "extraction" is missing a "c" in these descriptions

File: package.json [hunk 2]
```diff
@@ -80,7 +81,7 @@
     "prepublish": "npm run build",
     "release": "npmpub",
     "lint": "eslint . --ignore-path .gitignore && remark . --quiet --frail",
-    "tape": "babel-tape-runner \"src/**/__tests__/*.js\"",
+    "tape": "babel-tape-runner \"src/**/__tests__/index.js\"",
```
Comment (davidtheclark) [inline package.json]: Why is this change necessary?

File: src/__tests__/standalone-test.js (hunk 1 above)
Comment (davidtheclark) [inline src/__tests__/standalone-test.js]: Can we clarify this description? Am I right that it could be "... extractStyleTagsFromHtml option with straight CSS input"?

File: src/utils/extract.js [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+  }
+}
+
+function iterateStyles(code, onStyle) {
+  let index = 0
+  let currentStyle = null
+
+  const parser = new htmlparser.Parser({
+    onopentag(name) {
+      // Test if current tag is a valid <script> tag.
```
Comment (davidtheclark) [inline src/utils/extract.js]: Should be `<style>` tag, right?

File: src/__tests__/standalone-test.js [hunk 4]
```diff
@@ -395,6 +416,50 @@ test("standalone using codeFilename and ignoreFiles with configBasedir", t => {
...
+    code: "<script>\n</script>\n\n<style>\na {}\n</style>",
+    config: configExtendingOne,
+    configBasedir: fixturesPath,
+  }).then(({ output, results }) => {
+    t.equal(typeof output, "string")
+    t.equal(results.length, 1)
+    t.equal(results[0].warnings.length, 1)
+    t.equal(results[0].warnings[0].rule, "block-no-empty")
+    t.equal(results[0].warnings[0].line, 5)
+    t.equal(results[0].warnings[0].column, 10)
```
Comment (davidtheclark) [inline src/__tests__/standalone-test.js]: Why is this column 10? Should it be column 3?

File: src/utils/extract.js [hunk 5]
```diff
@@ -0,0 +1,96 @@
+const htmlparser = require("htmlparser2")
```
Comment (davidtheclark) [inline src/utils/extract.js]: Let's give the file a much more descriptive name, like other util files.

File: src/utils/extract.js [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+
+function parseIndentDescriptor(indentDescriptor) {
```
Comment (davidtheclark) [inline src/utils/extract.js]: Can you provide a comment please explaining what this is for?

File: src/utils/extract.js [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+
+  iterateStyles(code, function (previousCode, stylecode) {
+
+    // Mark that we're inside a <style> a tag and push all new lines
+    // in between the last </style> tag and this <style> tag to preserve
+    // location information.
+    const newLines = previousCode.match(/\r\n|\n|\r/g)
+    if (newLines) {
+      resultCode += newLines.map(function (newLine) {
+        return "/* fake-line */" + newLine
```
Comment (davidtheclark) [inline src/utils/extract.js]: Won't the addition of comments to the CSS before it is linted possibly cause phantom linting errors?

File: src/utils/extract.js [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+  const map = []
+  let resultCode = ""
+  let lineNumber = 1
+
+  iterateStyles(code, function (previousCode, stylecode) {
+
+    // Mark that we're inside a <style> a tag and push all new lines
+    // in between the last </style> tag and this <style> tag to preserve
+    // location information.
+    const newLines = previousCode.match(/\r\n|\n|\r/g)
```
Comment (davidtheclark) [inline src/utils/extract.js]: Is it necessary to look for solitary `\r`? Isn't it always `\r\n` or just `\n`?

File: src/utils/extract.js [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+      indent = indentMatch ? indentMatch[1] : ""
+    }
+    else {
+      indent = indentDescriptor.spaces
+      if (indentDescriptor.relative) {
+        indent = currentStyleIndent + indent
+      }
+    }
+
+    resultCode += stylecode
```
Comment (davidtheclark) [inline src/utils/extract.js]: Please some comments explaining your intentions in this file. There a lot of regular expressions that seem like they might be redundant, but I'm having a hard time figuring out exactly what is supposed to be happening, therefore uncertain what suggestions to make.

**Comment by davidtheclark** [comment]
Unless @jeddy3 has an opinion I'd say `extractStyleTagsFromHtml` is the most accurate option name I can think of.

So to finish this we'd need to:
- attend to comments
- put the feature behind a `extractStyleTagsFromHtml` option
- add documentation

Sound right?

File: src/utils/extract.js (hunk 8 above)
Comment (Arcanemagus) [inline src/utils/extract.js]: Windows uses `\r\n`, Pre-OSX Macs used `\r`, everything else (including OSX Macs) uses `\n`.

**Comment by jeddy3** [comment]
> Unless @jeddy3 has an opinion I'd say extractStyleTagsFromHtml is the most accurate option name I can think of.

SGTM. It's a great feature btw! @blake-newman Thanks for the contribution :)

**Comment by blake-newman** [comment]
@davidtheclark  @jeddy3 

Ill get some time to revise this PR beginning of the week, I agree with most of the comments and ill attend to.

File: src/utils/extract.js (hunk 8 above)
Comment (davidtheclark) [inline src/utils/extract.js]: We haven't worrying about the old mac format. I'd say leave it out here, to stick with the convention, and if you think that's a bad idea raise an issue so we fix it everywhere at once.

**Comment by blake-newman** [comment]
@davidtheclark @jeddy3 

This is slightly more difficult to implement the 'option' of this. Due to having to merge all the configuration before being able to parse `code` to postcss. Looking into the best way to do this with postcss, which is delaying me to revise the PR. 

Will be able to revise once this main implementation detail is working.

File: package.json (hunk 2 above)
Comment (blake-newman) [inline package.json]: It's not haha, i was being a donut :D

File: src/utils/extract.js (hunk 3 above)
Comment (blake-newman) [inline src/utils/extract.js]: Yes O_o

File: src/utils/extract.js (hunk 7 above)
Comment (blake-newman) [inline src/utils/extract.js]: Could possibly depending on the rules set, will investigate a better way to do this.

File: src/utils/extract.js (hunk 8 above)
Comment (blake-newman) [inline src/utils/extract.js]: 👍

**Comment by gaurav21r** [comment]
@blake-newman THanks for the fantastic PR. We urgently require this since scoped styling in **Polymer** (styles are inside `<style>`) isn't being linted. 

Can you elicit your problem in greater detail, maybe we can help?

**Comment by blake-newman** [comment]
I am working on this to get is releasable today. I'm at work so will likely be tonight at some point... 

Issue is using the postcss API to use the merged configuration settings. In order to parse the code, if `extractStyleTagsFromHtml` option is turned on. Only way i can see a way to do this is as an main option block.

```
{
  files,
  code,
  codeFilename,
  config,
  configFile,
  configBasedir,
  configOverrides,
  syntax,
  formatter = "json",
  extractStyleTagsFromHtml,
} = {})
```

**Comment by davidtheclark** [comment]
That seems reasonable to me.

**Comment by blake-newman** [comment]
@davidtheclark Updated code and tests. Documentation should be fairly easy to do.

**Comment by blake-newman** [comment]
Added documentation aswell

File: docs/user-guide/cli.md [hunk 10]
```diff
@@ -38,6 +38,12 @@ stylelint "foo/**/*.scss" --syntax scss
 
 The above can be slightly altered to read Less or SugarSS syntax: `--syntax less`, `--syntax sugarss`.
 
+```shell
+stylelint "foo/**/*.html" --extract
+```
+
+Extract `<style>` tags from html and lint.
```
Comment (davidtheclark) [inline docs/user-guide/cli.md]: In the other examples, the description is above the code sample and ends with a colon.

File: docs/user-guide/faq.md [hunk 11]
```diff
@@ -34,6 +34,10 @@ There are example of using the PostCSS plugin within the [docs](/docs/user-guide
 
 stylelint can *parse* any the following non-standard syntaxes: SCSS, Less and SugarSS. Refer to the [docs](/docs/user-guide/css-processors.md#parsing-non-standard-syntax) on how to configure stylelint to parse one of these syntaxes.
 
+## How do I lint styles within `style` tags?
```
Comment (davidtheclark) [inline docs/user-guide/faq.md]: `<style>` tags

File: docs/user-guide/faq.md [hunk 12]
```diff
@@ -34,6 +34,10 @@ There are example of using the PostCSS plugin within the [docs](/docs/user-guide
...
+
+stylelint can *extract and lint* any styles within `<style>` tags. Refer to the [docs](/docs/user-guide/node-api.md#extractStyleTagsFromHtml) for turning on this feature. You can also use the cli's `--extract` option to turn this feature on.
```
Comment (davidtheclark) [inline docs/user-guide/faq.md]: How about for the second sentence: "This can be done via the Node API and the CLI." (with links)

File: docs/user-guide/node-api.md [hunk 13]
```diff
@@ -67,6 +67,10 @@ Options: `"scss"|"less"|"sugarss"`.
 
 Specify a non-standard syntax that should be used to parse source stylesheets.
 
+### `extractStyleTagsFromHtml`
+
+Boolean. Enable extraction of css from inline `<style>` tags from within html structures, Keeping mapping to columns and lines.
```
Comment (davidtheclark) [inline docs/user-guide/node-api.md]: How about: "Enable extraction of CSS from `<style>` tags in HTML structures."

File: src/cli.js [hunk 14]
```diff
@@ -53,6 +53,7 @@ const meowOptions = {
     "                      (ignore \"warning\"-level)",
     "  -s, --syntax        Specify a non-standard syntax that should be used to ",
     "                      parse source stylesheets. Options: \"scss\", \"less\", \"sugarss\"",
+    "  -e, --extract       Sets extractStyleTagsFromHtml to true",
```
Comment (davidtheclark) [inline src/cli.js]: The other descriptions explain the option here instead of referring to other docs, so I'd prefer a short summary here, like "Extract and lint CSS from <style> tags in HTML structures"

File: src/cli.js [hunk 15]
```diff
@@ -80,6 +81,10 @@ if (cli.flags.syntax && includes(syntaxOptions, cli.flags.syntax)) {
   optionsBase.syntax = cli.flags.syntax
 }
 
+if (cli.flags.html) {
```
Comment (davidtheclark) [inline src/cli.js]: Shouldn't this flag be `extract`?

File: src/standalone.js [hunk 16]
```diff
@@ -94,6 +96,12 @@ export default function ({
     if (filepath) {
       postcssProcessOptions.from = filepath
     }
+    let extractMap = []
```
Comment (davidtheclark) [inline src/standalone.js]: What does is an empty array for this variable?

File: src/utils/styleTagsFromHtmlExtracter.js [hunk 17]
```diff
@@ -0,0 +1,83 @@
+import htmlparser from "htmlparser2"
+
+/**
+ * Extract styles from html markup, returning the exctacted
+ * code and mappings.
+ *
```
Comment (davidtheclark) [inline src/utils/styleTagsFromHtmlExtracter.js]: What should it return if the input `code` is not HTML? Or if there is no `<style>` tag in the HTML?

File: src/utils/styleTagsFromHtmlExtracter.js [hunk 18]
```diff
@@ -0,0 +1,83 @@
...
+ * @param {string} code
+ * @return {object} { map, code }
+ */
+export default function styleTagsFromHtmlExtracter(code) {
```
Comment (davidtheclark) [inline src/utils/styleTagsFromHtmlExtracter.js]: I think it's a great idea to abstract out this function, and I think it needs to have some unit tests to ensure it functions exactly as expected, and to clarify its functionality for people like me who are trying to figure out what it's supposed to do in a variety of situations.

**Comment by davidtheclark** [comment]
Cool, @blake-newman. I added some more comments. Because this is a pretty big feature, I expect there will be a fair bit of back-and-forth before we merge it in.

**Comment by blake-newman** [comment]
@davidtheclark i'll amend, and add the unit tests :)

File: src/utils/styleTagsFromHtmlExtracter.js (hunk 18 above)
Comment (blake-newman) [inline src/utils/styleTagsFromHtmlExtracter.js]: By abstraction, do you mean more comments or splitting into sub functions.

File: src/utils/styleTagsFromHtmlExtracter.js (hunk 17 above)
Comment (blake-newman) [inline src/utils/styleTagsFromHtmlExtracter.js]: Solved and added tests for scenario. If no extraction is done then it returns the original code input.

File: src/utils/__tests__/styleTagsFromHtmlExtracter-test.js [hunk 19]
```diff
@@ -0,0 +1,100 @@
...
+  // Expect result.code to be extracted code and a string
+  t.equal(typeof result.code, "string")
+  t.equal(result.code, expected)
+
+  // Expect result.map to be an Array with 0 mapped lines
+  t.equal(result.map instanceof Array, true)
+  t.equal(result.map.length, 0)
+
+  t.end()
+})
```
Comment (davidtheclark) [inline src/utils/__tests__/styleTagsFromHtmlExtracter-test.js]: Awesome tests. Super helpful.

**Comment by davidtheclark** [comment]
Looks great @blake-newman! 

@jeddy3 I'm fine with merging whenever you are.

**Comment by blake-newman** [comment]
@davidtheclark If you get any issues raised on this feature, ping me and ill try to assist :)

**Comment by jeddy3** [comment]
This is really nice! @blake-newman thanks for contributing the feature. @davidtheclark thanks for the careful reviewing. I've nothing else to add :)

I did a few basic smoke tests and everything worked great. The only thing I spotted missing during the tests was the `-e` alias, which I added in https://github.com/stylelint/stylelint/commit/0367b476366dcebd4d2b38c9726588742edd4688.

**Comment by gaurav21r** [comment]
Thanks @blake-newman @davidtheclark Now we can get **100%** linitng on all our styles :smile:
